### PR TITLE
relval update to fix hlt in 2017 wfs 

### PR DESCRIPTION
**Pull request by hengne**
- relval update to fix hlt in 2017 wfs :  
     add customization function as the default, to customize the hlt modules using the 2017 pixel geometry

- add one fullsim pu workflow ttbarlepton for hlt validation

- update gen-sim input string for 2017 workflows

### CODE REVIEW

**Comment by hengne** [comment]
please test

**Comment by hengne** [comment]
please test with PR #16867

**Comment by hengne** [comment]
please test with cms-sw/cmsdist#<16867>

**Comment by hengne** [comment]
please test with #16867

File: Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py [hunk 1]
```diff
@@ -87,13 +87,15 @@
         'Geom' : 'DB:Extended',
         'GT' : 'auto:phase1_2017_realistic',
         'HLTmenu': '@relval2016',
+        'Custom': 'HLTrigger/Configuration/customizeHLTTrackingForPhaseI2017.customizeHLTPhaseIPixelGeom',
```
Comment (makortel) [inline Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py]: AFAICT the customize gets now called for each of the workflow steps (not just HLT), but in practice the customize should have no effect on the other steps (in RECO `ClusterShapeHitFilterESProducer` `PixelShapeFile` already points to the "new" shape file).

Thought to note anyway.

**Comment by davidlange6** [comment]
> On Dec 6, 2016, at 6:17 AM, Matti Kortelainen <notifications@github.com> wrote:
> 
> @makortel commented on this pull request.
> 
> In Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:
> 
> > @@ -87,13 +87,15 @@
>          'Geom' : 'DB:Extended',
>          'GT' : 'auto:phase1_2017_realistic',
>          'HLTmenu': '@relval2016',
> +        'Custom': 'HLTrigger/Configuration/customizeHLTTrackingForPhaseI2017.customizeHLTPhaseIPixelGeom',
> 
> AFAICT the customize gets now called for each of the workflow steps (not just HLT), but in practice the customize should have no effect on the other steps (in RECO ClusterShapeHitFilterESProducer PixelShapeFile already points to the "new" shape file).
> 
> Thought to note anyway.

Right - but only because we insisted that the customize be implemented that way during the code review..

as for this request - the customize should just get called within the HLT (which already has such an infrastructure and not here. Apparently the goals have changed since the PR for the customization was made, as at that time it was first an alternative config meant to be cross validated and then made the default...

**Comment by makortel** [comment]
Adding @mtosi

**Comment by hengne** [comment]
it would be anyway best to implement this default in the hlt customization function, instead of being called here. seems to me it is possible to be implemented in the hlt GRun config file. no?

**Comment by hengne** [comment]
some thing like, at the end of 
HLTrigger/Configuration/python/HLT_GRun_cff.py

add: 
```
from Configuration.Eras.Modifier_phase1Pixel_cff import phase1Pixel

for esproducer in esproducers_by_type(fragment,"ClusterShapeHitFilterESProducer"):
    phase1Pixel.toModify(esproducer,  PixelShapeFile = 'RecoPixelVertexing/PixelLowPtUtilities/data/pixelShape_Phase1TkNewFPix.par'  )

for producer in producers_by_type(fragment,"SiPixelRawToDigi"):
      phase1Pixel.toModify(producer, UsePhase1=True)
```

**Comment by mtosi** [comment]
I would have done the same, indeed
but the bosses are @Martin-Grunewald  @silviodonato ...

**Comment by silviodonato** [comment]
Are you suggesting to create a new HLT_GRunFor2017 that is basically HLT_GRun_cff + 2017 customizer ? In this way, you want to call directly the HLT_GRunFor2017 in 2017 relvals, is it correct?

**Comment by hengne** [comment]
please test

**Comment by hengne** [comment]
+1

**Comment by hengne** [comment]
@davidlange6 if this hlt default implementation with eras in HLTrigger/Configuration still takes time, this PR need to be merged to temporarily make the workflows works (reasonable physics results) for pre2. although it should be applied in only the HLT/DIGI step, but the functionality to split steps for customization functions in the upgrade framework is still not yet implemented. considering this is only a temporary fix for the hlt default, and also considering the customization function will indeed not changing anything else than the hlt modules, seems to me this fix is an economic solution temporarily. I will remove it once the default is implemented with eras in HLTrigger packages.

**Comment by hengne** [comment]
the above comments also applies to the 81x PR #16882

**Comment by davidlange6** [comment]
Hi - @hengne @silviodonato - what is the down side to having a separate menu as @silviodonato suggests? Seems like the robust solution..

**Comment by Martin-Grunewald** [comment]
It needs to be maintained!

**Comment by davidlange6** [comment]
err - yes? we are planning to take data with this menu next year, right? I take it the HLT plan for 2016->17 differs from that of 2015->16. Is there a summary I can use as a guidance for whats to come?

**Comment by Martin-Grunewald** [comment]
Well, David, obviously no menu we have right now is the one we use in 2017.
This whole saga is just to use a current menu with some specific 2017 approximation.
If customisations for cmsDriver are no longer allowed, then lets drop these wflows
or make the matrix more flexible. The solution can not be to duplicate a menu and
then make a small customisation on the copy.

For what concerns the plan, please see the 1st/2nd page of
https://docs.google.com/document/d/1nbaiz_SvnqBs4UTKBZHrTMNYv9aNxU-3rAYu_QaucKo/edit?usp=sharing

**Comment by davidlange6** [comment]
ok, so yes, the plan is different than 2015-2016. Thanks. I would suggest putting your customize function into a new “menu” that simply wraps the 2016 one. that is by definition the same maintenance burden as a customize function.

**Comment by Martin-Grunewald** [comment]
Sorry, I do not see the point to add a menu which is the same as some other menu and adds some custom. function.

**Comment by davidlange6** [comment]
you presumably want two different configurations (with and without the proposed customize) to be runnable or no?

**Comment by Martin-Grunewald** [comment]
Yes - 2016 wflows with 2016 menus, 2017 wflows with 2016 menus plus custom - for now.
In the future, 2016 wflows (if still needed) with "fake" menus, and 2017 wflows with 2017 menus.

**Comment by fwyzard** [comment]
ok, github keeps dropping me out of the thread, so I'll have to go through it all...

in the meantime, are you suggesting to include a new file, e.g.
`HLTrigger/Configuration/python/HLT_GRun_2017_cff.py`
which includes something line
```
from HLTrigger.Configuration.HLT_GRun_cff import fragment
from HLTrigger.Configuration.customizeHLTTrackingForPhaseI2017 import customizeHLTPhaseIPixelGeom, customizeHLTForPFTrackingPhaseI2017

fragment = customizeHLTPhaseIPixelGeom(fragment)
fragment = customizeHLTForPFTrackingPhaseI2017(fragment)
```
?

The objection I have to this approach is that we try to maintain multiple triggers
  - one or to frozen menus
  - the GRun menu with the ongoing HLT development
  - the PbPb, p-Pb, and pp reference menus
  - etc
meaning we would need a similar file for each of them.

And, we should start putting together a menu for Phase 2, with again a similar set of customisations.
I would like to avoid keeping around 3 sets of 6 almost identical customisation files.

About Eras: I've been working on it last Friday and during the weekend, but the problem is that the current implementation of Eras is honestly quite crappy.
To me it would make much more sense if Eras were attached to the objects, not applied immediately.
More importantly, I need a consistent syntax for adding new objects that works with a full process and with a cff fragment.

**Comment by davidlange6** [comment]
Right, so the point is you have two configs you want to use. there are two non-hacky approaches. A new menu (which can be a few lines btw..), or eras. Both can do what you need

**Comment by fwyzard** [comment]
Except that Eras cannot.

**Comment by davidlange6** [comment]
I guess I missed what they can not do as it would be use case not brought up in the three years since proposed  - I’ll go back and find that thread - anyway, that leaves another option that lets the hlt control and maintain hlt configs in its area (which I think we can agree is the most appropriate solution…)

**Comment by fwyzard** [comment]
if you can tell me how to add a module to a cff file within an era, I'll be happy to use it to era-ify the HLT customisation.

**Comment by davidlange6** [comment]
I guess you mean to add a module to a sequence in an era?

On Dec 12, 2016, at 6:11 PM, Andrea Bocci <notifications@github.com<mailto:notifications@github.com>> wrote:


if you can tell me how to add a module to a cff file within an era, I'll be happy to use it to era-ify the HLT customisation.

**Comment by fwyzard** [comment]
Actually, I meant how to add a completely new module to the Process.

I do also need to add it to a Sequence, but that I can implement replacing the sequence with a modified copy of itself.

**Comment by fwyzard** [comment]
so, it looks like I found a way to do this with Eras, though I might be abusing the mechanism...

basically, you can do
```
eras.myEra.toModify(process, myCustomisationFunction)
```
to wrap the existing customisation functions inside `myEra`

In #16978 I've simply wrapped the two customise calls.
In #16977 I've tried to do it with a much finer granularity, but I don't know if that's actually better - it does look more confusing to me.

**Comment by fwyzard** [comment]
with both approaches

```python
import FWCore.ParameterSet.Config as cms
from Configuration.StandardSequences.Eras import eras

process = cms.Process( "HLT", eras.Run2_2017 )
process.load( 'HLTrigger.Configuration.HLT_GRun_cff' )
```

gives the same python configuration as

```python
process = cms.Process( "HLT" )
process.load( 'HLTrigger.Configuration.HLT_GRun_cff' )

from HLTrigger.Configuration.customizeHLTTrackingForPhaseI2017 import customizeHLTPhaseIPixelGeom
process = customizeHLTPhaseIPixelGeom(process)

from HLTrigger.Configuration.customizeHLTTrackingForPhaseI2017 import customizeHLTForPFTrackingPhaseI2017
process = customizeHLTForPFTrackingPhaseI2017(process)
```

**Comment by Martin-Grunewald** [comment]
Isn't that then always applied, regardless of which cmsDriver cmdline Era is chosen?

**Comment by fwyzard** [comment]
no, what I meant is that the customisation is applied only if you do 
```
process = cms.Process( "HLT", eras.Run2_2017 )
```
but not if you do
```
process = cms.Process( "HLT" )
```

**Comment by Martin-Grunewald** [comment]
Hmm, but where is the check on 2017 era - as opposed to 2016 era?
Ie, wouldn't the custom be applied also if you do ```process = cms.Process( "HLT", eras.Run2_2016 )```?
(2016 instead 2017)? IOW, where is the check or the connection to the 2017 era rather than to another era?

**Comment by fwyzard** [comment]
The customisations are attached to the `trackingPhase1` era:
```python
    # modify the HLT configuration for the Phase I pixel geometry
    from HLTrigger.Configuration.customizeHLTTrackingForPhaseI2017 import customizeHLTPhaseIPixelGeom
    eras.trackingPhase1.toModify(fragment, customizeHLTPhaseIPixelGeom)
```
etc.
 `trackingPhase1` is included in the `Run2_2017` era (see [Configuration/Eras/python/Era_Run2_2017_cff.py](https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/Configuration/Eras/python/Era_Run2_2017_cff.py) ), and not in the `Run2_2016` era (see [Configuration/Eras/python/Era_Run2_2016_cff.py](https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/Configuration/Eras/python/Era_Run2_2016_cff.py) )

**Comment by davidlange6** [comment]
Hi @hengne -this is now obsolete I think, is that correct?

**Pull request closed without merging**